docs/subnets/hello-world-precompile-tutorial.md

```diff
@@ -0,0 +1,979 @@
+
+```go
+ContractDeployerAllowListAddress = common.HexToAddress("0x0200000000000000000000000000000000000000")
+ContractNativeMinterAddress      = common.HexToAddress("0x0200000000000000000000000000000000000001")
+TxAllowListAddress               = common.HexToAddress("0x0200000000000000000000000000000000000002")
+FeeConfigManagerAddress          = common.HexToAddress("0x0200000000000000000000000000000000000003")
+HelloWorldAddress                = common.HexToAddress("0x0300000000000000000000000000000000000000")
+// ADD YOUR PRECOMPILE HERE
+// {YourPrecompile}Address       = common.HexToAddress("0x03000000000000000000000000000000000000??")
+```
+
+We now have to add it to the slice of `UsedAddresses` as well.
+
+```go
+UsedAddresses = []common.Address{
+		ContractDeployerAllowListAddress,
+		ContractNativeMinterAddress,
+		TxAllowListAddress,
+		FeeConfigManagerAddress,
+		HelloWorldAddress,
+		// ADD YOUR PRECOMPILE HERE
+		// YourPrecompileAddress
+	}
+```
+
+Now when Subnet-EVM sees the `HelloWorldAddress` as input when executing [`CALL`](https://github.com/ava-labs/subnet-evm/blob/master/core/vm/evm.go#L222), [`STATICCALL`](https://github.com/ava-labs/subnet-evm/blob/master/core/vm/evm.go#L401), [`DELEGATECALL`](https://github.com/ava-labs/subnet-evm/blob/master/core/vm/evm.go#L362), [`CALLCODE`](https://github.com/ava-labs/subnet-evm/blob/master/core/vm/evm.go#L311), it can [run the precompile](https://github.com/ava-labs/subnet-evm/blob/master/core/vm/evm.go#L271-L272) if the precompile is enabled.
+
+### Step 2: Set Gas Costs
+
+In `precompile/params.go` we have `writeGasCostPerSlot` and `readGasCostPerSlot`.
+
+`writeGasCostPerSlot` is the cost of one write such as modifying a state storage slot.
+
+`readGasCostPerSlot` is the cost of reading a state storage slot.
+
+This should be in your gas cost estimations based on how many times the precompile function does a read or a write. For example, if the precompile modifies the state slot of its precompile address twice then the gas cost for that function would be `40_000`.
+
+```go
+// Gas costs for stateful precompiles
+const (
+	writeGasCostPerSlot = 20_000
+	readGasCostPerSlot  = 5_000
+)
+```
+
+Now going back to `./precompile/hello_world.go`, we can modify our precompile function gas costs. Please search (`CTRL F`) `SET A GAS COST HERE` to locate the default gas cost code.
+
+We will be getting and setting our greeting with `sayHello()` and `setGreeting()` in one slot respectively so we can define the gas costs as follows.
+
+```go
+	SayHelloGasCost    uint64 = readGasCostPerSlot
+	SetGreetingGasGost uint64 = writeGasCostPerSlot
+```
+
+### Step 3: Add Custom Code
+
+Search (`CTRL F`) throughout the file with `CUSTOM CODE STARTS HERE` to find the areas in the precompile that we need to modify. We should start with the reference imports code block.
+
+![Custom code starts here](./../../static/img/2022-09-01-22-48-26.png)
```

Comment:
again if not very much required can we use a link to github or inline code?

Reply:
This shows the success state from running the script. No link possible. I would like to keep it in here unless there are strong opinions.